core/src/main/scala/org/apache/spark/SparkException.scala

```diff
@@ -355,3 +355,24 @@ private[spark] class SparkSQLFeatureNotSupportedException(
 
   override def getErrorClass: String = errorClass
 }
+
+/**
+ * User error exception thrown from Spark with an error class.
+ */
+private[spark] class SparkUserException(
```

Comment:
Question: Are we sure a custom exception is needed for this case? Is there any existing exceptions we can reuse with NPE as cause?

If we want to have a brand new exception, what about `SparkNotNullConstraintViolationException` to be more specific? I guess it will depend whether we want to skip retries only for this exception type as opposed to all Spark exceptions with known error codes.

Reply:
It's too complicated to use both error class and exception type to differentiate errors. I think in principle we should always use `SparkException` with different error classes, except for some places that need to be compatible with old code.

Reply:
There can be thousands of user-triggered errors while the transient errors are likely to be less than 10. I think it's better to define a new exception for transient errors.

Reply:
> I try to change the code according to this idea.

I see, thank you. I try to change the code according to this idea.